### ISSUE TITLE
Get magnitude of vector field as a field

### ISSUE THREAD

**Issue by tuckerbabcock** [mfem/mfem#2775]
Hi all, I'm wondering about possible approaches to use to get the magnitude of an RT vector field as a scalar field (probably in L2)?

The real use case for this is to get field that represents the magnitude of the magnetic flux density field. Right now I solve a magnetostatics problem to get the magnetic vector potential (in an ND space), and then use a `DiscreteLinearOperator` with a `CurlInterpolator` to get the flux density field in the RT space. I'd then like to get the magnitude of the flux density field as a scalar L2 field.

It seems like this would require writing a new `MixedBilinearFormIntegrator` to go from RT to L2, but I'm really wondering if there is a way to do this without assembling a global matrix, and ideally if I could go from ND to L2 in one step, and thus avoid the need to assemble two global matrices.

Thanks in advance for any help you're able to offer me!

Tucker

**Comment by tuckerbabcock** [comment]
Just thinking on it a little more, I guess it wouldn't be correct to use a `MixedBilinearFormIntegrator` since the process of going from RT to L2 (taking the magnitude) would be a nonlinear operation. Is there any support for rectangular nonlinear operators in MFEM?

**Comment by tzanio** [comment]
Hi @tuckerbabcock,

This is already done in GLVis using essentially `GridFunction::ProjectVectorFieldOn()`.

Specifically, you can replace the original vector grid function with (exact) L2 projection as in https://github.com/GLVis/glvis/blob/master/lib/stream_reader.cpp#L338-L362.

Then you can extract the components as L2 grid functions and compute the magnitude, see https://github.com/GLVis/glvis/blob/master/lib/vsvector3d.cpp#L267-L320.

Hope this helps,
Tzanio

**Comment by tuckerbabcock** [comment]
Hi @tzanio, thanks for pointing out those links. After reading through the linked code I've written my own implementation that lets me go from an ND state vector to a scalar DG field that holds the magnitude of the state vector's curl.

That seems to work great, but now I'm interested in reverse mode differentiating the output with respect to the input. I've gotten this working on one processor, but when I try to scale to multiple MPI processes my derivatives are no longer correct. Given the differentiation works in serial, I suspect the issue has something to do with correctly accounting for the sensitivity of the output to the ND dofs that are along the processor boundaries (since these dofs would impact DG dofs on either side of the boundary). I've been treating the assembled sensitivity as a "dual" vector, and thus using the `t_vec = P^T * l_vec` to assemble local contributions into the true dof vector, but clearly this isn't working for me. Is there something else I should be doing to assemble the local sensitivity contributions?

Thanks,

Tucker

**Comment by tuckerbabcock** [comment]
To give a little more detail, here I've visualized the sensitivity of the DG field output to an H1 field input that I've calculated with reverse mode AD (so contracted with an adjoint on the left had side). The first picture is with one process, then two, then four, and it is clear that there is some issue along the process boundaries. In the two and four process cases, rank 0 always seems to be visualized correctly, but the other ranks have an issue.
![1_proc_p2](https://user-images.githubusercontent.com/19217934/151471181-69151833-dec0-4ed0-869f-1a4e52577cf7.png)
![2_proc_p2](https://user-images.githubusercontent.com/19217934/151471182-70e94000-07dc-46f4-8e1f-37ec4611dc92.png)
![4_proc_p2](https://user-images.githubusercontent.com/19217934/151471185-62fe910e-c3ee-40b3-b846-aafdb6abba67.png)

Note that I'm visualizing these grid functions AFTER `grid_func.ParallelAssemble(true_vec)`, to the true dof vector, and then doing `fes->GetRestrictionOperator()->MultTranspose(true_vec, grid_func)` to reset the local dofs from the true ones. I suspect that the the `ParallelAssemble` part is the issue, since when I'm testing the derivatives for accuracy I only look at the true dof vectors, so the second step to get the local dofs is only used to visualize the field.

Please let me know of any suggestions for how to get this working correctly!

Thanks as always,

Tucker

**Comment by tzanio** [comment]
Hi @tuckerbabcock,

I am glad you got the magnitude of the curl working.

For the sensitivity, maybe @jandrej can help. With DG we sometimes need to call `ExchangeFaceNbrData`. Or maybe it is an ND orientation issue, have you visualized the produced curl field in parallel?

Tzanio

**Comment by tuckerbabcock** [comment]
Thanks for the reply @tzanio!

> have you visualized the produced curl field in parallel?

If you just mean just the forward part of the process, yes the DG curl field looks perfect in parallel. It is a little harder to visualize the sensitivity wrt the ND dofs though, since if I tried to print it to a Paraview file the sensitivity field would get interpolated with the ND shape functions and become a vector field which doesn't make much sense. For that reason, the screen shots I posted above are the sensitivity for just a dead simple H1 to L2 transformation, so then the sensitivities make more sense to plot interpolated by the H1 shape functions.

Since the issue is still there for the simple H1 to L2 transformation, I think there must be at least something else wrong besides any ND orientation issues (though there could still be these too for the real case I'm interested in).

I've not used `ExchangeFaceNbrData` before, so I think I may need a little assistance on how/when to use it.

Thank you again,

Tucker!

**Comment by tzanio** [comment]
In parallel DG we have a "ghost layer" of elements on the boundary of each domain, specifically the face-neighbors of elements that have a boundary face.

The DG values of these elements are updated with `ExchangeFaceNbrData` and are used in certain FEM operations, e.g. assembly.

**Comment by YohannDudouit** [comment]
If this is again a bug created by `ExchangeFaceNbrData` I think we should really consider calling this method in the constructor more automatically for users. I had many bugs due to an omitted call, or a delayed call to this method, I think @camierjs had a lot of bugs due to this too, just last month I helped @samuelpmishLLNL and @dylan-copeland debug their code due to omitted call to this method.
Wouldn't it be possible to simply call `ExchangeFaceNbrData` in `Assemble` or `Finalize` when `a->GetFBFI()->Size() > 0` or `a->GetBFBFI()->Size() > 0`?

**Comment by tuckerbabcock** [comment]
I'm not sure that `ExchangeFaceNbrData` is the issue here, since the field I'm visualizing and have shown above is an H1 dual field, and it seems that's what's not being assembled correctly. In fact for the reverse mode part, I don't even need the real DG output field, just a DG adjoint (primal) vector, which I've projected a coefficient onto. Would I need to call `ExchangeFaceNbrData` after projecting a coefficient onto the DG adjoint field?

**Comment by tuckerbabcock** [comment]
@jandrej do you have any thoughts for how I might be able to fix this?
